xapian-core/include/xapian/diversify.h

```diff
@@ -0,0 +1,134 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#ifndef XAPIAN_INCLUDED_DIVERSIFY_H
+#define XAPIAN_INCLUDED_DIVERSIFY_H
+
+#if !defined XAPIAN_IN_XAPIAN_H && !defined XAPIAN_LIB_BUILD
+#error "Never use <xapian/diversify.h> directly; include <xapian.h> instead."
+#endif
+
+#include <xapian/attributes.h>
+#include <xapian/cluster.h>
+#include <xapian/mset.h>
+#include <xapian/types.h>
+#include <xapian/visibility.h>
+
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+namespace Xapian {
+
+/** Class for diversifying an MSet using GLS-MPT as given
+ *  in the paper: Scalable and Efficient Web Search Result
+ *  Diversification Naini et al. 2016
+ */
+class XAPIAN_VISIBILITY_DEFAULT Diversify {
+    /// Top-k documents of given mset are diversified
+    unsigned int k;
+
+    /// MPT parameters
+    double lambda, b, sigma_sqr;
+
+    /// Store each document from given mset as a point
+    std::unordered_map<Xapian::docid, Xapian::Point> points;
+
+    /// Store the relevance score of each document
+    std::unordered_map<Xapian::docid, double> scores;
+
+    /// Store pairwise cosine similarities of documents of given mset
+    std::map<std::pair<Xapian::docid, Xapian::docid>, double> pairwise_sim;
+
+    /// Store docids of top k diversified documents
+    std::vector<Xapian::docid> main_dmset;
+
+    /** Initial diversified document set
+     *
+     *  Return top-k documents of mset as vector of Points, which
+     *  represent the initial diversified document set.
+     *
+     *  @param source	MSet object containing the documents of which
+     *			top-k are to be diversified
+     */
+    void initialise_points(const Xapian::MSet& source);
+
+    /** Return a key for a pair of documents
+     *
+     *  Returns a key as a pair of given documents ids
+     *
+     *  @param source	MSet object containing the documents of which
+     *			top-k are to be diversified
+     */
+    std::pair<Xapian::docid, Xapian::docid>
+    get_key(const Xapian::docid& docid_a, const Xapian::docid& docid_b);
+
+    /** Compute pairwise similarities
+     *
+     *  Used for pre-computing pairwise cosine similarities of documents
+     *  of given mset, which is used to speed up evaluate_dmset
+     */
+    void compute_similarities();
+
+    /** Return difference of 'points' and current dmset
+     *
+     *  Return the difference of 'points' and the current diversified
+     *  document match set
+     *
+     *  @param dmset	Document set representing a diversified document set
+     */
+    std::vector<Xapian::docid>
+    compute_diff_dmset(const std::vector<Xapian::docid>& dmset);
+
+    /** Evaluate a diversified mset
+     *
+     *  Evaluate a diversified mset using MPT algorithm
+     *
+     *  @param dmset	Set of points representing candidate diversifed
+     *			set of documents
+     */
+    double evaluate_dmset(const std::vector<Xapian::docid>& dmset);
```

Comment:
I'm happy to leave this to a follow-up PR, but I think that `Xapian::Diversify` should use the same PIMPL idiom which most of the Xapian API classes do - the currently exceptions are classes with a virtual method which the user can subclass (functor classes such as `Xapian::RangeProcessor`) and some classes which are just inherently cheap to copy and for which the implementation is fairly simple and so unlikely to be something we'd want to change during a release series (e.g. `Xapian::MSetIterator`).

It's surprising to have a new API class which is expensive to copy when the existing ones aren't (and the documentation clearly says they're cheap to copy).  Also, exposing the implementation ties our hands when it comes to changing the implementation during each stable release series, and also means more exported symbols from the shared library (which increases the load time cost), and potentially more STL headers implicitly included by `#include <xapian.h>`).

Reply:
This did cross my mind but being my first such PR, I didn't want to handle the added complexity which in hindsight I'm not sure was the right decision. But I don't think implementing the PIMPL idiom in this case would be much of a hassle now.